internal/controller/servermaintenance_controller.go

```diff
@@ -348,22 +422,295 @@ func (r *ServerMaintenanceReconciler) cleanup(ctx context.Context, log logr.Logg
+			status := &metalv1alpha1.ServerMaintenanceBootOrder{
+				State: metalv1alpha1.BootOrderConfigNoOp,
+			}
+			if err := r.patchBootOrderStatus(ctx, log, maintenance, status); err != nil {
+				return false, fmt.Errorf("failed to patch maintenance default boot order: %w", err)
+			}
+			// end state, continue with maintenance
+			return false, nil
+		}
+		bootOptions, err := bmcClient.GetBootOptions(ctx, server.Spec.SystemURI)
+		if err != nil {
+			return false, fmt.Errorf("failed to get boot options: %w", err)
+		}
+
+		var isDiskBootOption = func(bootOption *redfish.BootOption) bool {
+			diskIndicators := []string{"/HD(", "/Sata(", "/NVMe(", "/Scsi(", "/USB("}
+			for _, indicator := range diskIndicators {
+				if strings.Contains(bootOption.UefiDevicePath, indicator) {
+					return true
+				}
+			}
+			return false
+		}
+
+		var isPxeBootOption = func(bootOption *redfish.BootOption) bool {
+			return strings.Contains(strings.ToLower(bootOption.DisplayName), "pxe")
+		}
+
+		bootOrderChanged := rearrangeBootOrder(bootOrder, bootOptions, isPxeBootOption, isDiskBootOption)
+
+		if reflect.DeepEqual(bootOrder, bootOrderChanged) {
+			// boot order is already set to pxe first, nothing to do
+			log.V(1).Info("boot order for server is set to pxe first",
+				"Server", server.Name,
+				"BootOrder", bootOrder)
+			status := &metalv1alpha1.ServerMaintenanceBootOrder{}
+			if maintenance.Status.BootOrderStatus == nil {
+				status = &metalv1alpha1.ServerMaintenanceBootOrder{
+					State: metalv1alpha1.BootOrderConfigNoOp,
+				}
+			} else {
+				status = maintenance.Status.BootOrderStatus.DeepCopy()
+				status.State = metalv1alpha1.BootOrderConfigSuccess
+			}
+			if err := r.patchBootOrderStatus(ctx, log, maintenance, status); err != nil {
+				return false, fmt.Errorf("failed to patch maintenance default boot order: %w", err)
+			}
+			// end state, continue with maintenance
+			return false, nil
+		}
+
+		if maintenance.Status.BootOrderStatus == nil {
+			// note in this option, the server will always boot into pxe until changed again
+			// the job needs to be completed, else any other BIOS config job triggered will fail as exsisting job is not completed
+			if err := bmcClient.SetBootOrder(ctx, server.Spec.SystemURI, redfish.Boot{BootOrder: bootOrderChanged}); err != nil {
+				return false, fmt.Errorf("failed to set PXE boot once: %w", err)
+			}
+			log.V(1).Info("Configured boot order for server",
+				"Server", server.Name,
+				"BootType", maintenance.Spec.ServerBootConfigurationTemplate.BootType,
+				"BootOrder", bootOrderChanged)
+			// save the current boot order to revert back after maintenance
+			status := &metalv1alpha1.ServerMaintenanceBootOrder{
+				DefaultBootOrder: bootOrder,
+				State:            metalv1alpha1.BootOrderConfigInProgress,
+			}
+			if err := r.patchBootOrderStatus(ctx, log, maintenance, status); err != nil {
```

Comment:
Should `patchBootOrderStatus` be called before`SetBootOrder` in case patching the status fails? `SetBootOrder` would be called multiple times then.